### PR TITLE
Fix async usage

### PR DESCRIPTION
**Pull request by Omegastick**
It looks like async still isn't working for openai==1.1.0. This PR fixes all the issues I noticed:

- `apatch()` now patches the right method (and has been marked as unnecessary since it's the same method as `patch()`.
- A circular import that occured when you imported `apatch` has been fixed.
- Fixed a type error when retrying model response parsing.
- Fixed an OpenAI API error (`tool_calls` has to be either used or omitted) when retrying response parsing.
- Fixed an issue where the async OpenAI API object wasn't being detected as such, and was being patched as if it were synchronous.

I've added tests for those I could test, but the codebase isn't super testable so I didn't go too far out of my way.

## Summary by CodeRabbit

- **Refactor**
  - Reorganized import statements across various modules for better code organization.
  - Updated inheritance and import paths to enhance code maintainability.

- **New Features**
  - Introduced a `dump_message` function to improve message handling.
  - Added an `is_async` utility function to determine the nature of callables.

- **Bug Fixes**
  - Modified `retry_async` and `retry_sync` functions for improved error handling and consistency.

- **Tests**
  - Added new tests to ensure the correct functioning of the `patch`, `apatch`, `is_async`, and `wrap_chatcompletion` functions.
  - Enhanced existing tests with additional assertions for better coverage.

- **Documentation**
  - No visible changes to end-user documentation in this pull request.

### CODE REVIEW

File: tests/test_patch.py [hunk 1]
```diff
@@ -52,3 +56,57 @@ def validate_name(cls, v):
...
     ), "The raw response should be available from OpenAI"
+
+
+def test_patch_completes_successfully():
+    instructor.patch(OpenAI())
+
+
+def test_apatch_completes_successfully():
+    instructor.apatch(AsyncOpenAI())
+
```
[the remark this answers is not in the document]
Comment (Omegastick) [reply tests/test_patch.py:67]: Not much of an idea on what to assert here. Open to suggestions. This is mainly just checking for calls for missing methods, etc.

File: instructor/dsl/validators.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
-import openai
-from pydantic import Field
 from typing import Optional
+
 from openai import OpenAI
-import instructor
+from pydantic import Field
```
[the remark this answers is not in the document]
Comment (Omegastick) [reply instructor/dsl/validators.py:4]: These are both used in the file.

**Comment by jxnl** [comment]
1) would love some tips on making it more testable
2) does this overlap with #165 ? 

looking now

**Comment by jxnl** [comment]
Noticed some failing tests on union types. Can take a look tomorrow!

**Comment by pablopalafox** [comment]
Hi @jxnl and @Omegastick! `examples/validators/allm_validators.py` would also need fixing, I think.

**Pull request merged**